### PR TITLE
feat(spanner): add interval type support

### PR DESCRIPTION
**Pull request by rahul2393**

### CODE REVIEW

File: spanner/value.go [hunk 1]
```diff
@@ -5065,3 +5214,340 @@ func trimDoubleQuotes(payload []byte) ([]byte, error) {
...
+	n.Valid = true
+	return nil
+}
+
+// ParseInterval parses an ISO8601 duration format string into an Interval.
+func ParseInterval(s string) (Interval, error) {
+	var interval Interval
+	interval.Nanos = new(big.Int)
+
+	pattern := `^P(-?\d+Y)?(-?\d+M)?(-?\d+D)?(T(-?\d+H)?(-?\d+M)?(-?(\d+(\.\d{1,9})?|\.\d{1,9})S)?)?$`
```
Comment (sagarwaal) [inline spanner/value.go:5380]: decimal can be represented through both `,` and '.'

File: spanner/integration_test.go [hunk 2]
```diff
@@ -858,6 +880,195 @@ func TestIntegration_SingleUse_WithQueryOptions(t *testing.T) {
...
+
+	for i := range intervals {
+		if intervals[i].Valid != wantIntervals[i].Valid || intervals[i].Interval.Months != wantIntervals[i].Interval.Months ||
+			intervals[i].Interval.Days != wantIntervals[i].Interval.Days ||
+			(intervals[i].Interval.Nanos != nil && wantIntervals[i].Interval.Nanos != nil && intervals[i].Interval.Nanos.Cmp(wantIntervals[i].Interval.Nanos) != 0) {
+			t.Errorf("interval %d: got %+v, want %+v", i, intervals[i], wantIntervals[i])
+		}
+	}
+}
+
```
Comment (sagarwaal) [inline spanner/integration_test.go:1071]: 1. Can we also have tests for passing parameters of type Interval and Interval Array (both NULL and non NULL params).
2. Can we also have tests for reading NULL Interval and NULL Interval Array.

File: spanner/value.go [hunk 3]
```diff
@@ -5065,3 +5214,345 @@ func trimDoubleQuotes(payload []byte) ([]byte, error) {
...
+
+	if i.Nanos != nil && i.Nanos.Sign() != 0 {
+		result.WriteString("T")
+		nanos := new(big.Int).Set(i.Nanos)
+		isNegative := nanos.Sign() < 0
+		if isNegative {
+			nanos.Neg(nanos)
+		}
+
+		nanosPerHour := new(big.Int).SetInt64(3600000000000)
```
Comment (sagarwaal) [inline spanner/value.go:5276]: should we use constants, same at the other places?

File: spanner/value.go [hunk 4]
```diff
@@ -5065,3 +5214,345 @@ func trimDoubleQuotes(payload []byte) ([]byte, error) {
...
+
+			secondsNanos := new(big.Int).SetInt64(secondsInt * 1000000000)
+			interval.Nanos.Add(interval.Nanos, secondsNanos)
+
+			if fractionStr != "" {
+				for len(fractionStr) < 9 {
+					fractionStr += "0"
+				}
+				if len(fractionStr) > 9 {
+					fractionStr = fractionStr[:9]
```
Comment (sagarwaal) [inline spanner/value.go:5469]: we should never reach this, the regex parsing will fail for more than 9 digits after decimal.

**Review by sagarwaal — APPROVED** [review]

**Review by harshachinta — APPROVED** [review]
